packages/core/src/modules/ledger/services/IndyPoolService.ts

```diff
@@ -0,0 +1,154 @@
+   */
+  public async getPoolForDid(did: string): Promise<{ pool: IndyPool; did: Indy.GetNymResponse }> {
+    const pools = this.pools
+
+    if (pools.length === 0) {
+      throw new LedgerNotConfiguredError(
+        "No indy ledgers configured. Provide at least one pool configuration in the 'indyLedgers' agent configuration"
+      )
+    }
+
+    const { successful, rejected } = await this.getSettledDidResponsesFromPools(did, pools)
+
+    // This means there is a rejected response besides not found.
+    // For now we're really strict, and if one request fails, the whole
+    // request will fail. We may loosen this is the future if we see fit
+    const rejectedOtherThanNotFound = rejected.filter((e) => !(e.reason instanceof LedgerNotFoundError))
+    if (rejectedOtherThanNotFound.length > 0) {
+      throw new LedgerError(
+        `Unknown error retrieving did '${did}' from '${rejectedOtherThanNotFound.length}' of '${pools.length}' ledgers`,
+        { cause: rejectedOtherThanNotFound[0].reason }
+      )
+    }
+
+    // If the successful is empty at this stage, it means all ledgers did not have the did
+    if (successful.length === 0) {
+      throw new LedgerNotFoundError(`Did '${did}' not found on any of the ledgers (total ${this.pools.length}).`)
+    }
+
+    // Split between production and nonProduction ledgers. If there is at least one
+    // successful response from a production ledger, only keep production ledgers
+    // otherwise we only keep the non production ledgers.
+    const production = successful.filter((s) => s.value.pool.config.isProduction)
+    const nonProduction = successful.filter((s) => !s.value.pool.config.isProduction)
+    const remaining = production.length >= 1 ? production : nonProduction
+
+    // If there's only one successful response in this group (production / nonProduction), return it
+    if (remaining.length === 1) {
+      return { pool: remaining[0].value.pool, did: remaining[0].value.did }
+    }
+
+    // If there are multiple successful responses in this group check only one of them is
+    // self-certifying. Return if this is the case
+    const selfCertifying = remaining.filter((response) =>
+      isSelfCertifiedDid(response.value.did.did, response.value.did.verkey)
+    )
+    if (selfCertifying.length === 1) {
+      return { pool: selfCertifying[0].value.pool, did: selfCertifying[0].value.did }
+    }
+
+    // FIXME: ACA-Py doc mentions txnTime, but doc from Stephen mentions that it is not
+    // possible to get when the did was created (it is updated when the did is rotated probably)
+    const sortedByTxnTime = [...remaining].sort(
+      (first, second) =>
+        (first.value.response as Indy.LedgerReadReplyResponse).result.txnTime -
+        (second.value.response as Indy.LedgerReadReplyResponse).result.txnTime
+    )
+    return { pool: sortedByTxnTime[0].value.pool, did: sortedByTxnTime[0].value.did }
+  }
```

Comment:
This is the filtering mechanism as described in the PR body

Reply:
Awesome stuff!

Two comments.  First -- the second check about an error.  I'm assuming that the error is not on one ledger, but on all of the ledgers -- is that correct?  That is -- if one ledger succeeds, and another errors off, I think we would just continue?

Next, on the ordering on the last point about txn times.  I agree that we don't want a service to track when DIDs are created.  But we also don't want to throw an error if at least some ledgers were resolved, so some ordering needs to be used to return the result.  We know we have either all production, or all non-production ledgers returned.  Could the order of the configuration list be used?

Alternatively -- is it worth passing the list to the controller?  Seems like a lot of work, but in theory the controller could select the one to be used.  I'm thinking it's probably not worth it, but thought I would raise it.

Reply:
No, currently if one fails (with an unexpected error) we won't continue. Let's say sovrin main is down and has the did. Is that same did from BCovrin also valid then? But thinking about it the other way around, lets say BCovrin is down, should retrieving something from any other ledger also fail then? No, of course not. I'll loosen it up.

Yes, I think the ordering of the pools could be used here. This also gives some control to the user to decide which networks to choose.

It may be worth passing the list to the controller, but this is just shifting the problem, and probably something 90% of users won't need/don't want. I think using the order of the list as you mentioned already gives some control.


Thanks for the feedback 🙂